fix(AppFramework): Correctly implement CSRF checks

provokateurin opened the pull request:
## Summary

These changes ensure that actually all the CSRF protection methods can be correctly used.
I improved the existing codes and allowed requests with OCS-APIRequest headers to pass which was not possible before and a bug.
As far as I can see there were no security problems before and this does not open any new ones :tm:

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

nickvergessen reviewed (changes requested):
Seems to break cypress?

ChristophWurst commented on lib/private/AppFramework/Http/Request.php:
```diff
@@ -451,47 +451,31 @@ protected function decodeContent() {
-		$token = new CsrfToken($token);
-
-		return $this->csrfTokenManager->isTokenValid($token);
+		return $this->getHeader('OCS-APIREQUEST') || $this->passesStrictCookieCheck() || $this->passesCSRFTokenCheck();
```
this means the check passes if strict cookies are sent but no or an invalid no CSRF token. should it be `&&`?

provokateurin replied:
From my understanding either of the methods are fine, so this should be correct.

provokateurin replied:
I started reading some literature about this topic and you are right.
By allowing the request if it passes the cookie check an evil actor that is already on your server could send you a compromised link (e.g. as a Talk message) and it would execute just fine because you are already on the correct host and send the cookies.

To my understanding the cookie check is actually not doing anything at all and doesn't give us any additional protection, because it will not protect against the attack mentioned above and the other methods already provide sufficient protection (especially since very old browsers (7+ years) treat same-site cookies as normal cookies).

The header is a sufficient way to protect against CSRF attacks because there is no way for an attacker to set it in a malicious request.

ChristophWurst replied:
Makes sense!

come-nc commented:
Code looks good but I do not know the situation well enough to say if it’s okay to require only one method to pass.

provokateurin committed:
refactor(AppFramework): Improve strict and lax cookie checks

Signed-off-by: provokateurin <kate@provokateurin.de>

provokateurin committed:
refactor(AppFramework): Improve and document cookie check requirements

Signed-off-by: provokateurin <kate@provokateurin.de>

provokateurin committed:
fix(AppFramework): Allow requests with OCS-APIRequest header to pass CSRF checks

Signed-off-by: provokateurin <kate@provokateurin.de>

provokateurin committed:
docs(AppFramework): Document CSRF check requirements

Signed-off-by: provokateurin <kate@provokateurin.de>

provokateurin commented:
I rewrote this PR from scratch with the new information I had. This should be much easier to review. The main problem before was concerning the cookie checks which I didn't understand correctly, but now I also added documentation for the next people who will look at it...

come-nc reviewed (commented):
In the end what is the behavior change brought by this PR?

provokateurin commented:
It allows using the OCS-APIRequest header to bypass CSRF checks. For clients this is necessary, so they don't have to deal with CSRF tokens. Both methods are safe, but the former was not possible until now.

nickvergessen commented:
So nothing for 29 during freeze :speak_no_evil:

provokateurin commented:
I guess so :/

provokateurin commented:
Since the branch-off is done, can I get reviews to get it into Nextcloud 30?

provokateurin commented:
@nickvergessen @ChristophWurst Can I get your reviews?

Altahrim commented on lib/private/AppFramework/Http/Request.php:
```diff
@@ -451,8 +451,10 @@ protected function decodeContent() {
+	 * Checks if the request passes the CSRF checks.
+	 *
+	 * A request must always pass the strict cookie check, unless it has the OCS-APIRequest set or no session (@see cookieCheckRequired).
+	 * If the OCS-APIRequest is set or a valid CSRF token is sent the request will to pass.
```
```suggestion
	 * If the OCS-APIRequest is set or a valid CSRF token is sent the check will succeed.
```

Altahrim commented on lib/private/AppFramework/Http/Request.php:
```diff
@@ -463,6 +465,10 @@ public function passesCSRFCheck(): bool {
 			return false;
 		}
 
+		if ($this->getHeader('OCS-APIRequest') !== '') {
+			return true;
+		}
```
I think we can move this condition before the `passesStrictCookieCheck`. It saves a few calls.

provokateurin replied:
I agree, although I think it makes sense to always validate the cookies. If invalid cookies are sent they should be rejected regardless of the header.

Altahrim replied:
According to the *current* code, `passesStrictCookieCheck` basically return `true` if `OCS-APIRequest`  is not empty. It saves two function call.
Your solution is more future proof ^^

Altahrim commented on lib/private/AppFramework/Http/Request.php:
```diff
@@ -481,17 +487,12 @@ public function passesCSRFCheck(): bool {
 	/**
 	 * Whether the cookie checks are required
 	 *
+	 * In case the OCS-APIRequest header is set or the user has no session, we don't to check the cookies because the client is not a browser and thus doesn't need CSRF checks.
```
```suggestion
	 * In case the OCS-APIRequest header is set or the user has no session, we don't need to check the cookies because the client is not a browser and thus doesn't need CSRF checks.
```

provokateurin committed:
fixup! docs(AppFramework): Document CSRF check requirements

provokateurin committed:
fixup! refactor(AppFramework): Improve and document cookie check requirements

nickvergessen commented on tests/lib/AppFramework/Http/RequestTest.php:
```diff
@@ -2219,7 +2219,7 @@ public function invalidTokenDataProvider() {
 	 * @param string $invalidToken
 	 */
-	public function testPassesCSRFCheckWithInvalidToken($invalidToken) {
+	public function testFailsCSRFCheckWithInvalidToken($invalidToken) {
```
We usually try to follow the pattern `test...MethodName...ScenarioDescription()`, since the method to test is called `passesCSRFCheck()` the old name matches the pattern, but I also see how it can be confusing in this case. The problem is the actual method name, but we should also not change that... So maybe just stick with the old name of test cases.

provokateurin replied:
Ah I see, I thought the `passes` was from the it passes/fails and not from the method name. I agree that the method name is not good :+1:

provokateurin commented:
I made a minimal version of these changes that only include the actual fix and leaves out all the refactoring and documentation: https://github.com/nextcloud/server/pull/46760
